### ISSUE TITLE
[BUG]: Typing.h not working with gcc-14 (line 104)

### ISSUE THREAD

**Issue by 97gamjak** [pybind/pybind11#5206]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### What version (or hash if on master) of pybind11 are you using?

v2.13.0 & v2.13.1 & #5203

### Problem description

So basically, with tag v 2.13.0 a bug considering the GNU 14 compiler was introduced in typing.h line 104 with std::copy_n

tag v2.12.0 works.

So looking at the changelog of the GNU compiler they state [this](https://gcc.gnu.org/gcc-14/porting_to.html) on how to port the code to gcc-14. 

"""
C++ language issues
Header dependency changes
Some C++ Standard Library headers have been changed to no longer include other headers that were being used internally by the library. As such, C++ programs that used standard library components without including the right headers will no longer compile.

The following headers are used less widely in libstdc++ and may need to be included explicitly when compiling with GCC 14:

<algorithm> (for std::copy_n, std::find_if, std::lower_bound, std::remove, std::reverse, std::sort etc.)
...
"""

### Is this a regression? Put the last known working version here if it is.

v2.12.0

**Comment by 97gamjak** [comment]
Yes, but shouldn't this `#include` be included in the typing.h file?

**Comment by Mizux** [comment]
seems a duplicate of:
* #5201 

TLDR: so version of gcc seems to miss some c++20 STL template functions...

**Comment by rwgk** [comment]
> Yes, but shouldn't this #include be included in the typing.h file?

This was a pybind11 bug (missing include) that was fixed already on master. It'll be in the next release.

```
$ git show 51c2aa16de5b50fe4be6a0016d6090d4a831899e
```
```diff
commit 51c2aa16de5b50fe4be6a0016d6090d4a831899e
Author: wenqing <wenqing.wang@ufz.de>
Date:   Fri Jun 28 16:12:32 2024 +0200

    Fixed a compilation error with gcc 14 (#5208)

diff --git a/include/pybind11/typing.h b/include/pybind11/typing.h
index c8ba18d4..b0feb946 100644
--- a/include/pybind11/typing.h
+++ b/include/pybind11/typing.h
@@ -14,6 +14,8 @@
 #include "cast.h"
 #include "pytypes.h"

+#include <algorithm>
+
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
 PYBIND11_NAMESPACE_BEGIN(typing)
```
